src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -216,8 +228,8 @@ desugarCase (Case scrut alternatives) =
           alt_fail = [CaseAlternative [NullBinder] [MkUnguarded goto_rem_case]]
 
         pure $ Let [
-            ValueDeclaration rem_case_id Private [NullBinder]
-              [MkUnguarded desugared]
+          ValueDeclaration rem_case_id Private []
+            [MkUnguarded (Abs (VarBinder unused_binder) desugared)]
```

Comment:
This was the culprit. After desugaring we need to generate code which obeys the invariants of desugared code..